server/remote_asset/fetch_server/fetch_server.go

```diff
@@ -203,11 +192,80 @@ func (p *FetchServer) FetchDirectory(ctx context.Context, req *rapb.FetchDirecto
 	return nil, status.UnimplementedError("FetchDirectory is not yet implemented")
 }
 
+func (p *FetchServer) rewriteToCache(ctx context.Context, blobDigest *repb.Digest, instanceName string, fromFunc, toFunc repb.DigestFunction_Value) *repb.Digest {
+	cacheRN := digest.NewResourceName(blobDigest, instanceName, rspb.CacheType_CAS, fromFunc)
+	cache := p.env.GetCache()
+	reader, err := cache.Reader(ctx, cacheRN.ToProto(), 0, 0)
+	if err != nil {
+		log.CtxErrorf(ctx, "Failed to get cache reader for %s: %s", digest.String(blobDigest), err)
+		return nil
+	}
+
+	tmpFilePath, err := tempCopy(reader)
+	if err != nil {
+		log.CtxErrorf(ctx, "Failed to copy from reader to temp for %s: %s", digest.String(blobDigest), err)
+		return nil
+	}
+	defer func() {
+		if err := os.Remove(tmpFilePath); err != nil {
+			log.Errorf("Failed to remove temp file: %s", err)
+		}
+	}()
+
+	bsClient := p.env.GetByteStreamClient()
+	storageDigest, err := cachetools.UploadFile(ctx, bsClient, instanceName, toFunc, tmpFilePath)
+	if err != nil {
+		log.CtxErrorf(ctx, "Failed to re-upload blob with new digestFunc %s for %s: %s", toFunc, digest.String(blobDigest), err)
+		return nil
+	}
+	return storageDigest
+}
+
+func (p *FetchServer) findBlobInCache(ctx context.Context, instanceName string, checksumFunc repb.DigestFunction_Value, expectedChecksum string) *repb.Digest {
+	blobDigest := &repb.Digest{
+		Hash: expectedChecksum,
+		// The digest size is unknown since the client only sends up
+		// the hash. We can look up the size using the Metadata API,
+		// which looks up only using the hash, so the size we pass here
+		// doesn't matter.
+		SizeBytes: 1,
+	}
+	cacheRN := digest.NewResourceName(blobDigest, instanceName, rspb.CacheType_CAS, checksumFunc)
+	log.CtxInfof(ctx, "Looking up %s in cache", blobDigest.Hash)
```

Comment:
wdyt about removing these log.Info lines in the non-error cases? I think they might be a tad excessive

Reply:
Good suggestion.

I will switch these to Debug level instead of removing them.
These could be fairly useful for development purposes.